### PR TITLE
Coral-trino: Migrate Alias appending transformer from RelNode To SqlNode layer

### PR DESCRIPTION
**Pull request by aastha25**
Currently, `RelToTrinoConverter` hosts multiple transformations required to convert a CoralRelNode to a Trino-compatible SQL statement. This PR intends to relocate one such transformation from RelNode layer to SqlNode layer. 

Currently, AS operator is inserted while accessing nested fields. For ex: `SELECT structField.intField FROM foo` is converted to `SELECT structField.intField AS intField FROM foo`. Previously, this transformation was done at the Rel->SqlNode layer. This PR moves the change to SqlNode rewritten layer. 

Testing:
Integ tested with all the production data sources. No regression was observed. 
Also refactored `RelToTrinoConverter` to utilize `HiveToRelConverter`  

**Note:** This PR captures a more minor change from a larger effort here https://github.com/linkedin/coral/pull/315  to test backward compatibility in smaller chunks. The difference in translations due to PR#315 is large. The changes have been split over multiple PRs to reduce the effort of testing and be able to monitor the impact on production views more closely.

### CODE REVIEW

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:30]: `SqlNodes`

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (yiqiangin) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:51]: declare as `private`?

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:33]: Please do not use `TrinoSqlNode` since it is not a class or well defined entity. In fact, if anything it should be the final output, not the input.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:34]: Stick to one in the entire comment (SqlNode or SqlCall).

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (yiqiangin) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:85]: Does Coral use Apache Commons Collection? If yes, I think it is better to call `Collections.isEmpty()` to check if the list is null or empty().

**Review by yiqiangin — APPROVED** [review]
LGTM

**Comment by ljfgem** [comment]
@aastha25 could you elaborate the reason why we need to separate this PR from #315 ? Would be great if we don't need to introduce a temp class `IntermediateTrinoSqlNodeRewriter` which will be modified soon.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:94]: `Always add`
Convert all SQL to uppercase.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:102]: Comment is not clear. Please contrast the behavior before and after override clearly.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java:85]: Tried changing to:
`if (CollectionUtils.isNotEmpty(((SqlSelect) sqlCall).getSelectList().getList()))`

it throws NPE since `getSelectList()` might be null. `getSelectList()` returns type `SqlNodeList` which would also need to be checked.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/TrinoSqlDialect.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/TrinoSqlDialect.java:77]: this public method was an override for https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/SqlDialect.java#L612 without the annotation. 

It should be safe to delete.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java]: done

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java]: updated the documentation. Please review it again when you get a chance.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/IntermediateTrinoSqlNodeRewriter.java]: updated the documentation. Please review it again when you get a chance.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java]: updated the documentation. Please review it again when you get a chance.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/SqlNodeConverter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/SqlNodeConverter.java:65]: Coral IR allows implicit casting to VARCHAR from other data types such as INTs, etc. However, Trino requires explicit casting.

**Review by wmoustafa — COMMENTED** [review]
LGTM after addressing one comment and ensuring regression tests pass.

**Comment by aastha25** [comment]
@wmoustafa the integ tests succeeded and I have addressed the open comment. Could you give it a final pass?

**Review by wmoustafa — APPROVED** [review]

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java
Comment (ljfgem) [inline coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:378]: I am worried about the query result regression if we add `try_cast` for some operators like `>`/`<`.
For example:
HQL:
```
SELECT * FROM t WHERE 10 > 6; // always true
```
Translated Trino SQL:
```
SELECT * FROM t WHERE try_cast(10, varchar) > try_cast(6, varchar); // always false because '10' < '6'
```
This kind of regression can't be caught by regression test, but will affect the query result.
cc: @wmoustafa

**Review by ljfgem — COMMENTED** [review]
LGTM except one concern.

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java
Comment (wmoustafa) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:378]: Good catch. The agreement was to fix this behavior once the type inference work is checked in. Now that it is checked in, maybe we can fix it as part of this PR.

Comment (aastha25) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java]: @ljfgem @wmoustafa  I have updated the PR to address this concern. PTAL

Comment (ljfgem) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java]: Thanks @aastha25, have you run the regression test again for this update? Let's ensure there is no translation failure.

**Comment by ljfgem** [comment]
@wmoustafa @aastha25 After investigation and tests, I found that we may not need `RelationalOperatorTransformer` (original [adjustInconsistentTypesToEqualityOperator](https://github.com/linkedin/coral/blob/master/coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/Calcite2TrinoUDFConverter.java#L426)) to begin with.

Reason:
For relational operation like `'1' = 1` where there is a type mismatch in CoralSqlNode1, Calcite will add explicit `CAST` in Coral IR automatically during `CoralSqlNode1 -> Coral IR` and the corresponding converted RexCall becomes 
`=(CAST('1'):INTEGER NOT NULL, 1)`, then CoralSqlNode2 becomes `CAST('1' AS INTEGER) = 1`. So we don't need to add `TRY_CAST` anymore (Actually this PR already rules it out [here](https://github.com/linkedin/coral/pull/329/files#diff-7f8a049f1695fd79362914170997ba47b067eb80d098f5ae01cedeb1dd88a5e3R76)).
Regression test also proved it - there is no added `TRY_CAST` in any production view with this PR.
I have also verified that there is no regression if we just remove [adjustInconsistentTypesToEqualityOperator](https://github.com/linkedin/coral/blob/master/coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/Calcite2TrinoUDFConverter.java#L426) without introducing `RelationalOperatorTransformer` with Trino local server.

In conclusion, I think we can  remove [adjustInconsistentTypesToEqualityOperator](https://github.com/linkedin/coral/blob/master/coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/Calcite2TrinoUDFConverter.java#L426) directly without introducing `RelationalOperatorTransformer` in SqlNode layer.

**Comment by ljfgem** [comment]
@aastha25 As we discussed before, we need to remove the unnecessary `RelationalOperatorTransformer`, so this PR doesn't contain any transformation migrations, then I think it would be great if you can split this PR to 2 PRs to keep PRs smaller and more specific:
1. PR that includes the work in `SqlNodeConverter`, which is related to the work in #315 
2. PR that includes the API refactoring and type derivation enhancement. Since these parts are for transformation migration with where type derivation is needed, I would suggest to add another transformation migration with type derivation (i.e. `visitConcat`) together with them, so that we can add the minimum code and test API refactoring & type derivation enhancement parts.

cc: @wmoustafa

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java:28]: Could you make the name descriptive what the transformation does?

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java:26]: Could you add an example and a justification for the transformation in the comment?

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (wmoustafa) [inline coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:114]: What makes this PR include this change? Could you explain the need for it in the description?

Comment (wmoustafa) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:114]: Also, should we assume transformation at this layer will be converted later to SqlNode level?

**Comment by wmoustafa** [comment]
Let us move the tranformation in #373 to SqlNode layer as well.

**Comment by aastha25** [comment]
> Let us move the tranformation in #373 to SqlNode layer as well.
@wmoustafa  we can include the corresponding SqlNode transformer - `RelationalOperatorTransformer` but it'll be a no-op given the existing translation path.

**Comment by wmoustafa** [comment]
> @wmoustafa we can include the corresponding SqlNode transformer - `RelationalOperatorTransformer` but it'll be a no-op given the existing translation path.

I think it is fine. At some point we should add tests that do not go through RelNodes.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:114]: I'll add an example to clarify the need for the override. 
At this point, we should expect this transformation to stay here in the Rel -> SqlNode layer. This transformation helps achieve parity between sqlNode1 & sqlNode2.

Comment (wmoustafa) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:114]: Will it then move to Coral RelNode to Coral SqlNode converter?

**Comment by ljfgem** [comment]
> Let us move the tranformation in #373 to SqlNode layer as well.

@wmoustafa If we want to ensure CoralSqlNode1 == CoralSqlNode2, I would prefer to add a transformation on `HiveSqlNode -> CoralSqlNode1` conversion path to add explicit `CAST` for type mismatch in HQL because:
1. Even if we move the transformation in #373 to SqlNode layer (`CoralSqlNode2 -> TrinoSqlNode`), CoralSqlNode1 != CoralSqlNode2 because Calcite adds the explicit `CAST`
2. Transformation in #373 just replaced `CAST` by `TRY_CAST`, which is not necessary for Trino SQL as I mentioned before:
```
If CAST succeeds, then CAST is equivalent to TRY_CAST.
If CAST fails, a query using CAST fails, while a query using TRY_CAST succeeds. However, since TRY_CAST returns null in this case, it may lead to query result data quality issues, which is undesirable.
```
And I would prefer to add/migrate transformations in another PR because #329 focuses on adding `SqlSelectListTransformer`, it would be great to keep PRs small and specific.

**Comment by wmoustafa** [comment]
@ljfgem It is still equivalent since the Coral IR spec in that case is that we are just saying it the sides of comparison operators do not necessarily match. If through the RelNode path the CAST was introduced, it is still acceptable. It is optional in the first place.
For `CAST` vs `TRY_CAST`, yes I prefer `CAST`.
I will let @aastha25 decide on which PR.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java]: done.

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/SqlSelectListTransformer.java]: I have added an example in the java doc. I have also updated the PR description to explain the motivation. Tldr: This is a migration from RelNode -> SqlNode

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java]: yes.

**Comment by aastha25** [comment]
> @ljfgem SqlNode1 and SqlNode2 are still equivalent since the Coral IR spec in that case is that we are just saying that the sides of comparison operators do not necessarily match. If through the RelNode path the CAST was introduced, it is still acceptable. It is optional in the first place. For `CAST` vs `TRY_CAST`, yes I prefer `CAST`. I will let @aastha25 decide on which PR.

I prefer to add `RelationalOperatorTransformer` it in a separate PR after #378 is merged. This is a transformer that needs data type derivation and will rely on the new API for type derivation.

**Review by ljfgem — COMMENTED** [review]
Thanks @aastha25 for this PR!

File: coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java
Comment (aastha25) [reply coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/RelToTrinoConverter.java:343]: this seems to be a part of the legacy code, that is not used. Removing it has no impact on translations.

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java
Comment (aastha25) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java]: renamed the function.

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java
Comment (aastha25) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java]: removed the exception

**Review by ljfgem — COMMENTED** [review]
LGTM except [a remaining comment](https://github.com/linkedin/coral/pull/329/files#r1156435615)
@wmoustafa @yiqiangin please also take a look if you get a chance.

File: coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java
Comment (ljfgem) [inline coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:790]: nit: a space after `//`, or we can remove this comment since there is no similar comment for other UTs.

Comment (aastha25) [reply coral-trino/src/test/java/com/linkedin/coral/trino/rel2trino/HiveToTrinoConverterTest.java:790]: I can fix it in the next PR if its not a blocking concern. Thanks

**Pull request merged**
